### PR TITLE
Gradient cutting by local comparisons

### PR DESCRIPTION
**Pull request by gobbleturk**
This helps with unstable training (large training loss spikes), however it is by no means a complete fix. There are still several spikes seen with this change, but we have seen better loss over time, with fewer spikes that are recovered from faster.

### CODE REVIEW

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -109,3 +109,9 @@ prompt: "I love to "
 enable_profiler: False
 enable_checkpointing: True
 
+grad_cut_factor: 1.2 # Gradient capped by mean of previous gradients times this factor.
```
Comment (rwitten) [inline MaxText/configs/base.yml:112]: `grad_cut` is a bad name unfortunately. My best is `rolling_gradient_clipping`.

File: MaxText/configs/base.yml [hunk 2]
```diff
@@ -109,3 +109,9 @@ prompt: "I love to "
...
+grad_cut_window_len: 100 # The window length of running gradients to compare to.
+grad_cut_start_step: 100 # Gradient starts being cut at this step.
```
Comment (rwitten) [inline MaxText/configs/base.yml:114]: Probably safe to assume these are the same for now?

File: MaxText/train.py [hunk 3]
```diff
@@ -203,13 +204,18 @@ def loss_fn(params):
 
   grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
   (loss, intermediate_outputs), grads = grad_fn(state.params)
+  pre_cut_grad_norm = jnp.array(max_utils.l2norm_pytree(grads))
+  max_grad_norm = jnp.array(config.grad_cut_factor * grad_cut)
+  grad_scale_factor = jnp.minimum(max_grad_norm, pre_cut_grad_norm) / pre_cut_grad_norm
+  grads = jax.tree_map(lambda g: g * grad_scale_factor, grads)
```
Comment (rwitten) [inline MaxText/train.py:210]: I think it is best to call this `scaled_grads`.

File: MaxText/train.py [hunk 4]
```diff
@@ -203,13 +204,18 @@ def loss_fn(params):
...
+  grad_norm = max_utils.l2norm_pytree(grads)
   new_state = state.apply_gradients(grads=grads)
-  metrics = {'scalar': {'learning/loss': loss, 'learning/grad_norm' : max_utils.l2norm_pytree(grads),
+  metrics = {'scalar': {'learning/loss': loss, 'learning/grad_norm' : grad_norm,
```
Comment (rwitten) [inline MaxText/train.py:213]: as metrics we probably want `raw grad norm`, `applied grad norm` and `rolling gradient clipping threshold`

File: MaxText/configs/base.yml (hunk 1 above)
Comment (rwitten) [reply MaxText/configs/base.yml:112]: So we'd call the various quantities lower in the code `rolling_gradient_clipping_window_len`, `rolling_gradient_clipping_start_step`, `rolling_gradient_clipping_scale_factor`, `rolling_gradient_clipping_current_threshold`, etc.

File: MaxText/train.py [hunk 5]
```diff
@@ -318,23 +324,31 @@ def train_loop(config, state=None):
...
   example_batch = None
   last_step_completion = datetime.datetime.now()
+  grad_norms = deque(maxlen=config.grad_cut_window_len)
+  LARGE_GRADIENT_NORM = 1.0 * 10**6 # Used when gradient cutting is not desired.
 
   local_metrics_file = open(config.metrics_file, 'a', encoding="utf8") if config.metrics_file else None
 
   for step in np.arange(get_first_step(state), config.steps):
     example_batch = load_next_batch(train_iter, example_batch, config)
+    if step < config.grad_cut_start_step or not grad_norms:
```
Comment (rwitten) [inline MaxText/train.py:343]: put this into its own function! and don't define `LARGE_GRADIENT_NORM`, I think `float('inf')` has the right semantics here.

**Review by rwitten — COMMENTED** [review]
Excited to see the results for this.
